File: 2024/Day_02/part02.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <string>
#include <sstream>
using namespace std;

void readData(vector<vector<int>> &data) {
    ifstream file("test.txt");
    if (!file.is_open()){
        cerr << "Error Opening File" << endl;
        exit(-1);
    }

    string line;
    while(getline(file, line)) {
        istringstream ss(line);
        int val;
        vector<int> row;
        while (ss >> val) row.push_back(val);
        data.push_back(row);
    }
}

bool isSafe(vector<int> row) {
    bool isIncrem = row[0] < row[1];
    for (int i = 1; i < row.size(); i++) {
        int dist = abs(row[i] - row[i-1]);
        if (dist < 1 || dist > 3) return false;
        if ((isIncrem && row[i] < row[i-1]) || (!isIncrem && row[i] > row[i-1]))
            return false;
    }
    return true;
}

int main() {
    vector<vector<int>> data;
    readData(data);
    int numSaves = 0;

    // checks each row id safe or not
    for(int i = 0; i < data.size(); i++) {
        // checks if row is safe if remove ith elemnent
        for(int j = 0; j < data[i].size(); j++) {
            vector<int> newRow = data[i];
            newRow.erase(newRow.begin()+j);
            if (isSafe(newRow)) {
                numSaves++; 
                break;
            }
        }
    }

    cout << numSaves << endl;
    return 0;
}

File: 2024/Day_02/part02.py
import numpy as np
import csv

def is_safe(row):
    is_increm = row[0] < row[1]
    for i in range(1, len(row)):
        dist = abs(row[i] - row[i-1])
        if dist < 1 or dist > 3: 
            return False
        if (is_increm and row[i] < row[i-1]) or (not is_increm and row[i] > row[i-1]):
            return False
    return True

# reads data into lists of lists
data = []
with open('test.txt', 'r') as file:
    reader = csv.reader(file, delimiter=' ')
    for row in reader:
        data.append([int(val) for val in row])

# checks each row id safe or not
num_saves = 0
for row in data:
    # checks if row is safe if remove ith elemnent
    for i in range(len(row)):
        new_row = row[:i] + row[i+1:]
        if is_safe(new_row): 
            num_saves += 1
            break

print(num_saves)
